third_party/nvfuser/csrc/scheduler/reduction_utils.cpp

```diff
@@ -511,6 +521,143 @@ void multiReductionInliner(
   inlineMost();
 }
 
+void propagateTransformation(
+    TensorView* reference_tv,
+    const std::unordered_set<TensorView*>& boundaryNodesSet) {
+  // set an internal boundary to allow different parallel patterns in the fusion
+  struct InternalBoundarySelector : public MaxInfoSpanningTree::Selector {
+    std::unordered_set<TensorView*> _tvs;
+    virtual bool allowC2P(TensorView* from, TensorView* to) override {
+      return _tvs.count(from) == 0;
+    };
+    virtual bool allowP2C(TensorView* from, TensorView* to) override {
+      return _tvs.count(to) == 0;
+    };
+    virtual bool allowSibling(TensorView* from, TensorView* to) override {
+      return true;
+    }
+    InternalBoundarySelector(const std::unordered_set<TensorView*>& tvs)
+        : _tvs(tvs) {}
+  };
+  InternalBoundarySelector ibSelector(boundaryNodesSet);
+  TransformPropagator propagator(reference_tv);
+  MaxRootDomainInfoSpanningTree(reference_tv, &ibSelector)
+      .traverse(&propagator);
+}
+
+void propagateRFactor(
```

Comment:
Can this be used to refactor `multiReductionInliner`?

Reply:
should be able to do it. Let's leave the refactor to another PR.